sdk/resource/resource.go

```diff
@@ -17,101 +17,118 @@
+	if r == nil {
+		r = Empty()
 	}
-	b.WriteRune(')')
-
-	return b.String()
+	return r.labels.Encoded(label.DefaultEncoder())
 }
 
 // Attributes returns a copy of attributes from the resource in a sorted order.
-func (r Resource) Attributes() []core.KeyValue {
-	return append([]core.KeyValue(nil), r.sorted...)
+// To avoid allocating a new slice, use an iterator.
+func (r *Resource) Attributes() []core.KeyValue {
+	if r == nil {
+		r = Empty()
+	}
+	return r.labels.ToSlice()
 }
 
 // Iter returns an interator of the Resource attributes.
-//
 // This is ideal to use if you do not want a copy of the attributes.
-func (r Resource) Iter() AttributeIterator {
-	return NewAttributeIterator(r.sorted)
+func (r *Resource) Iter() label.Iterator {
+	if r == nil {
+		r = Empty()
+	}
+	return r.labels.Iter()
 }
 
-// Equal returns true if other Resource is equal to r.
-func (r Resource) Equal(other Resource) bool {
-	return r.String() == other.String()
+// Equal returns true when a Resource is equivalent to this Resource.
+func (r *Resource) Equal(eq *Resource) bool {
+	if r == nil {
+		r = Empty()
+	}
+	if eq == nil {
+		eq = Empty()
+	}
+	return r.Equivalent() == eq.Equivalent()
 }
 
 // Merge creates a new resource by combining resource a and b.
 // If there are common key between resource a and b then value from resource a is preserved.
 // If one of the resources is nil then the other resource is returned without creating a new one.
 func Merge(a, b *Resource) *Resource {
 	if a == nil {
-		return b
+		a = Empty()
 	}
 	if b == nil {
-		return a
+		b = Empty()
 	}
+	// Note: 'b' is listed first so that 'a' will overwrite with
+	// last-value-wins in label.New()
+	combine := append(b.Attributes(), a.Attributes()...)
+	return New(combine...)
+}
 
-	// Note: the following could be optimized by implementing a dedicated merge sort.
+// Empty returns an instance of Resource with no attributes.  It is
+// equivalent to a `nil` Resource.
+func Empty() *Resource {
+	return &emptyResource
+}
 
-	kvs := make([]core.KeyValue, 0, len(a.sorted)+len(b.sorted))
-	kvs = append(kvs, a.sorted...)
-	// a overwrites b, so b needs to be at the end.
-	kvs = append(kvs, b.sorted...)
-	return New(kvs...)
+// Equivalent returns an object that can be compared for equality
+// between two resources.  This value is suitable for use as a key in
+// a map.
+func (r *Resource) Equivalent() label.Distinct {
+	if r == nil {
+		r = Empty()
+	}
+	return r.labels.Equivalent()
 }
 
-// MarshalJSON prints the resource attributes in sorted order.
-func (r Resource) MarshalJSON() ([]byte, error) {
-	return json.Marshal(r.sorted)
+// MarshalJSON encodes labels as a JSON list of { "Key": "...", "Value": ... }
+// pairs in order sorted by key.
+func (r *Resource) MarshalJSON() ([]byte, error) {
+	if r == nil {
+		r = Empty()
+	}
+	return r.labels.MarshalJSON()
+}
+
+// Len returns the number of unique key-values in this Resource.
+func (r *Resource) Len() int {
+	if r == nil {
+		r = Empty()
```

Comment:
nit:

```suggestion
		return 0
```

Minor, but would avoid the possible use of a reflect value. :shrug:

Reply:
Done. (There's no reflection in `Empty()`--not sure what you mean.)

Reply:
Ah, your are correct. I was mis-remembering to when unique values were stored with reflect Values and confusing the two. My bad.